File: week02/D-flowshop/main.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n, m;
    cin >> n >> m;
    vector p(n, vector(m, 0ll));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> p[i][j];
        }
    }
    vector<int> ans(n), wait(m);
    for (int j = 0; j < m; j++) {
        for (int i = 0; i < n; i++) {
            ans[i] = max(ans[i], wait[j]) + p[i][j];
            wait[j] = ans[i];
        }
    }
    for (const auto &e : ans) {
        cout << e << ' ';
    }
    cout << endl;
    return 0;
}

File: week02/D-flowshop/main.rs
use std::{cmp::max, io};

fn main() {
    let stdin = io::read_to_string(io::stdin()).unwrap();
    let mut stdin = stdin.split_whitespace();
    let n: usize = stdin.next().unwrap().parse().unwrap();
    let m: usize = stdin.next().unwrap().parse().unwrap();

    let p: Vec<Vec<i64>> = (0..n)
        .map(|_| {
            (0..m)
                .map(|_| stdin.next().unwrap().parse().unwrap())
                .collect()
        })
        .collect();

    let mut ans = vec![0; n];
    let mut wait = vec![0; m];

    for (j, w) in wait.iter_mut().enumerate() {
        for i in 0..n {
            ans[i] = max(ans[i], *w) + p[i][j];
            *w = ans[i];
        }
    }

    for e in &ans {
        print!("{} ", e);
    }
    println!()
}
